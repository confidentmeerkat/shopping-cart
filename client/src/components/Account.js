import React from 'react';
import moment from 'moment';
import numeral from 'numeral';
import { connect } from 'react-redux';
import { getUser } from '../actions/userActions';
import AccountModal from './AccountModal';
import Divider from 'material-ui/Divider';
import RaisedButton from 'material-ui/RaisedButton';
import Edit from 'material-ui/svg-icons/image/edit.js';
import '../styles/Account.css';

class Account extends React.Component {

  state = {
    accountModalOpen: false
  }

  toggleAccountModal = () => {
    this.setState({ accountModalOpen: !this.state.accountModalOpen });
  }

  componentDidMount() {
    this.props.getUser();
  }

  render() {
    if(!this.props.user) {
      return (
        <div className="account-container">
          <div className="loader">
            <img src="/img/loader.gif" />
            <h1>LOADING ACCOUNT DATA...</h1>
          </div>
        </div>
      );
    } else {
      return (
        <div className="account-container">
          <h1>Your Account</h1>
          <div className="account">
            <div className="account-info">
              <h2>Info</h2>
              <Divider />
              <p><b>Username: </b>{this.props.user.username}</p>
              <p><b>E-mail: </b>{this.props.user.email}</p>
              <p><b>Billing Address: </b>{this.props.user.address}</p>
              <p><b>Phone: </b>{this.props.user.phone}</p>
              <RaisedButton
                className="btn"
                label="Edit account"
                labelPosition="before"
                primary={true}
                icon={<Edit />}
                onClick={this.toggleAccountModal}
              />
            </div>
            <div className="account-history">
              <h2>Order History</h2>
              <Divider />
              <div class="orders">
                {this.props.user.orders.length ? 
                  <table>
                    <tr>
                      <th>Date Created</th>
                      <th>Product Name</th>
                      <th>Price</th>
                      <th>Qty</th>
                      <th>Total</th>
                    </tr>
                    {this.props.user.orders.map((order) => {
                      return (
                        <tr>
                          <td>{moment(order.dateCreated).format('ll')}</td>
                          <td>{order.name}</td>
                          <td>{numeral(order.price).format('$0,0.00')}</td>
                          <td>{order.quantity}</td>
                          <td>{numeral(order.price * order.quantity).format('$0,0.00')}</td>
                        </tr>
                      );
                    })}
                  </table> :
                  <h1>No order history.</h1>
                }
              </div>
            </div>
          </div>
          <AccountModal
            user={this.props.user}
            isOpen={this.state.accountModalOpen}
            onRequestClose={this.toggleAccountModal}
          />
        </div>
      );
    }
  }
};

const mapStateToProps = (state) => ({
  user: state.loggedUser
});

const mapDispatchToProps = (dispatch) => ({
  getUser: () => dispatch(getUser())
});

export default connect(mapStateToProps, mapDispatchToProps)(Account);