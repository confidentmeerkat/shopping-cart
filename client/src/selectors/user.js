export const selectUser = (state) => state.loggedUser.user;