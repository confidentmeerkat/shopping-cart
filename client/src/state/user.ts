import { Order } from './order';

export interface User {
  orders: Order[];
  username: string;
  email: string;
  address: string;
  phone: string;
  _id: string;
}
