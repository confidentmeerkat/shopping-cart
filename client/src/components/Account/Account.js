import React from 'react';
import moment from 'moment';
import numeral from 'numeral';
import AccountModal from '../AccountModal';
import Divider from 'material-ui/Divider';
import RaisedButton from 'material-ui/RaisedButton';
import Edit from 'material-ui/svg-icons/image/edit.js';
import '@styles/Account.css';

class Account extends React.Component {
  state = {
    accountModalOpen: false
  }

  toggleAccountModal = () => {
    this.setState({ accountModalOpen: !this.state.accountModalOpen });
  }

  componentDidMount() {
    this.props.getUser();
  }

  render() {
    const { user } = this.props;

    if(!user) {
      return (
        <div className="account-container">
          <div className="loader">
            <img src="/img/loader.gif" />
            <h1>LOADING ACCOUNT DATA...</h1>
          </div>
        </div>
      );
    } else {
      return (
        <div className="account-container">
          <h1>Your Account</h1>
          <div className="account">
            <div className="account-info">
              <h2>Info</h2>
              <Divider />
              <p><b>Username: </b>{user.username}</p>
              <p><b>E-mail: </b>{user.email}</p>
              <p><b>Billing Address: </b>{user.address}</p>
              <p><b>Phone: </b>{user.phone}</p>
              <RaisedButton
                className="btn"
                label="Edit account"
                labelPosition="before"
                primary={true}
                icon={<Edit />}
                onClick={this.toggleAccountModal}
              />
            </div>
            <div className="account-history">
              <h2>Order History</h2>
              <Divider />
              <div className="orders">
                {user.orders.length ?
                  <table>
                    <thead>
                      <tr>
                        <th>Date Created</th>
                        <th>Product Name</th>
                        <th>Price</th>
                        <th>Qty</th>
                        <th>Total</th>
                      </tr>
                    </thead>
                    <tbody>
                      {user.orders.map((order) => {
                        return (
                          <tr key={order.name}>
                            <td>{moment(order.dateCreated).format('ll')}</td>
                            <td>{order.name}</td>
                            <td>{numeral(order.price).format('$0,0.00')}</td>
                            <td>{order.quantity}</td>
                            <td>{numeral(order.price * order.quantity).format('$0,0.00')}</td>
                          </tr>
                        );
                      })}
                    </tbody>
                  </table> :
                  <h1>No order history.</h1>
                }
              </div>
            </div>
          </div>
          <AccountModal
            user={user}
            isOpen={this.state.accountModalOpen}
            onRequestClose={this.toggleAccountModal}
          />
        </div>
      );
    }
  }
};

export default Account;
