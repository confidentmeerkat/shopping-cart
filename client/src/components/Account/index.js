import { compose } from 'recompose';
import { connect } from 'react-redux';
import { getUser } from '@actions';
import Account from './Account';

const mapStateToProps = (state) => ({
  user: state.loggedUser.user
});

const actions = { getUser };

export default compose(
  connect(mapStateToProps, actions)
)(Account);
