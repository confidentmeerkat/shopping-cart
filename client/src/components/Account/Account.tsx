import * as React from 'react';
import * as moment from 'moment';
import * as numeral from 'numeral';
import { IUser, IOrder } from '@typings/state/index';
import IconButton from '@material-ui/core/IconButton';
import EditIcon from '@material-ui/icons/Edit';
import Divider from '@material-ui/core/Divider';
import AccountModal from '../AccountModal';
import '@styles/Account.css';

interface Props {
  user: IUser;
  getUser: () => any;
}

interface State {
  accountModalOpen: boolean;
}

class Account extends React.Component<Props, State> {
  state = {
    accountModalOpen: false
  }

  toggleAccountModal = () => {
    this.setState({ accountModalOpen: !this.state.accountModalOpen });
  }

  componentDidMount() {
    this.props.getUser();
  }

  render() {
    const { user } = this.props;

    if(!user) {
      return (
        <div className="account-container">
          <div className="loader">
            <img src="/img/loader.gif" />
            <h1>LOADING ACCOUNT DATA...</h1>
          </div>
        </div>
      );
    } else {
      return (
        <div className="account-container">
          <h1>Your Account</h1>
          <div className="account">
            <div className="account-info">
              <div className="top">
                <h2>Info</h2>
                <IconButton
                  color="secondary"
                  aria-label="Edit"
                  onClick={this.toggleAccountModal}>
                  <EditIcon />
                </IconButton>
              </div>
              <Divider />
              <p><b>Username: </b>{user.username}</p>
              <p><b>E-mail: </b>{user.email}</p>
              <p><b>Billing Address: </b>{user.address}</p>
              <p><b>Phone: </b>{user.phone}</p>
            </div>
            <div className="account-history">
              <h2>Order History</h2>
              <Divider />
              <div className="orders">
                {user.orders.length ?
                  <table>
                    <thead>
                      <tr>
                        <th>Date Created</th>
                        <th>Product Name</th>
                        <th>Price</th>
                        <th>Qty</th>
                        <th>Total</th>
                      </tr>
                    </thead>
                    <tbody>
                      {user.orders.map((order: IOrder) => {
                        return (
                          <tr key={order.name}>
                            <td>{moment(order.dateCreated).format('ll')}</td>
                            <td>{order.name}</td>
                            <td>{numeral(order.price).format('$0,0.00')}</td>
                            <td>{order.quantity}</td>
                            <td>{numeral(parseInt(order.price) * parseInt(order.quantity)).format('$0,0.00')}</td>
                          </tr>
                        );
                      })}
                    </tbody>
                  </table> :
                  <h1>No order history.</h1>
                }
              </div>
            </div>
          </div>
          <AccountModal
            user={user}
            isOpen={this.state.accountModalOpen}
            onRequestClose={this.toggleAccountModal}
          />
        </div>
      );
    }
  }
};

export default Account;
