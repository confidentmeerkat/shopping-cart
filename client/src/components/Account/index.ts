import { compose } from 'recompose';
import { connect } from 'react-redux';
import { getUser } from '@actions/index';
import { selectUser } from '@selectors/user';
import { IState } from '@typings/state/index';
import Account from './Account';

const mapStateToProps = (state: IState) => ({
  user: selectUser(state)
});

const actions = { getUser };

export default compose(
  connect(mapStateToProps, actions)
)(Account);
