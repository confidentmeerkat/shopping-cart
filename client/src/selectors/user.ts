import { IState, IUser } from '@typings/state/index';

export const selectUser = (state: IState): IUser | null => state.loggedUser.user;
