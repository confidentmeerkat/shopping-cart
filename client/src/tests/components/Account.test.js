import React from 'react';
import { shallow } from 'enzyme';
import { Account } from '../../components/Account';
import user from '../__mocks__/user';

describe('<Account />', () => {
  it('renders Account with no user logged correctly', () => {
    const wrapper = shallow(<Account user="" getUser={() => { }} />);

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('img').prop('src')).toBe('/img/loader.gif');
    expect(wrapper.find('h1').text()).toBe('LOADING ACCOUNT DATA...');
  });

  it('renders Account with user logged correctly', () => {
    const wrapper = shallow(<Account user={user} getUser={() => {}} />);

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('.account-info').children()).toHaveLength(7);
    expect(wrapper.find('.account-history').children()).toHaveLength(3);
    expect(wrapper.find('tbody').children()).toHaveLength(2);
  });

  it('opens modal on button click', () => {
    const wrapper = shallow(<Account user={user} getUser={() => { }} />);

    wrapper.find('RaisedButton').simulate('click');
    expect(wrapper.state('accountModalOpen')).toBe(true);
  });
});