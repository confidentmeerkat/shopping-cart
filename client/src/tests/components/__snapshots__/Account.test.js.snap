// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Account /> renders Account with no user logged correctly 1`] = `
<div
  className="account-container"
>
  <div
    className="loader"
  >
    <img
      src="/img/loader.gif"
    />
    <h1>
      LOADING ACCOUNT DATA...
    </h1>
  </div>
</div>
`;

exports[`<Account /> renders Account with user logged correctly 1`] = `
<div
  className="account-container"
>
  <h1>
    Your Account
  </h1>
  <div
    className="account"
  >
    <div
      className="account-info"
    >
      <h2>
        Info
      </h2>
      <Divider
        inset={false}
      />
      <p>
        <b>
          Username: 
        </b>
        John
      </p>
      <p>
        <b>
          E-mail: 
        </b>
        john@mail.com
      </p>
      <p>
        <b>
          Billing Address: 
        </b>
        1st street
      </p>
      <p>
        <b>
          Phone: 
        </b>
        123 456
      </p>
      <RaisedButton
        className="btn"
        disabled={false}
        fullWidth={false}
        icon={<ImageEdit />}
        label="Edit account"
        labelPosition="before"
        onClick={[Function]}
        primary={true}
        secondary={false}
      />
    </div>
    <div
      className="account-history"
    >
      <h2>
        Order History
      </h2>
      <Divider
        inset={false}
      />
      <div
        className="orders"
      >
        <table>
          <thead>
            <tr>
              <th>
                Date Created
              </th>
              <th>
                Product Name
              </th>
              <th>
                Price
              </th>
              <th>
                Qty
              </th>
              <th>
                Total
              </th>
            </tr>
          </thead>
          <tbody>
            <tr
              key="Apple iPhone"
            >
              <td>
                Feb 24, 2018
              </td>
              <td>
                Apple iPhone
              </td>
              <td>
                $700.00
              </td>
              <td>
                1
              </td>
              <td>
                $700.00
              </td>
            </tr>
            <tr
              key="Samsung Galaxy"
            >
              <td>
                Feb 24, 2018
              </td>
              <td>
                Samsung Galaxy
              </td>
              <td>
                $800.00
              </td>
              <td>
                1
              </td>
              <td>
                $800.00
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
  </div>
  <AccountModal
    isOpen={false}
    onRequestClose={[Function]}
    user={
      Object {
        "_id": "123",
        "address": "1st street",
        "email": "john@mail.com",
        "orders": Array [
          Object {
            "dateCreated": 1519512681240,
            "name": "Apple iPhone",
            "price": 700,
            "quantity": 1,
          },
          Object {
            "dateCreated": 1519512681240,
            "name": "Samsung Galaxy",
            "price": 800,
            "quantity": 1,
          },
        ],
        "phone": "123 456",
        "username": "John",
      }
    }
  />
</div>
`;
